File: csharp/Svix/Connector.cs
// this file is @generated
#nullable enable
using Microsoft.Extensions.Logging;
using Svix.Models;

namespace Svix
{
    public class ConnectorListOptions : SvixOptionsBase
    {
        public ulong? Limit { get; set; }
        public string? Iterator { get; set; }
        public Ordering? Order { get; set; }

        public new Dictionary<string, string> QueryParams()
        {
            return SerializeParams(
                new Dictionary<string, object?>
                {
                    { "limit", Limit },
                    { "iterator", Iterator },
                    { "order", Order },
                }
            );
        }
    }

    public class ConnectorCreateOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class Connector(SvixClient client)
    {
        readonly SvixClient _client = client;

        /// <summary>
        /// List all connectors for an application.
        /// </summary>
        public async Task<ListResponseConnectorOut> ListAsync(
            ConnectorListOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response =
                    await _client.SvixHttpClient.SendRequestAsync<ListResponseConnectorOut>(
                        method: HttpMethod.Get,
                        path: "/api/v1/connector",
                        queryParams: options?.QueryParams(),
                        headerParams: options?.HeaderParams(),
                        cancellationToken: cancellationToken
                    );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ListAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// List all connectors for an application.
        /// </summary>
        public ListResponseConnectorOut List(ConnectorListOptions? options = null)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ListResponseConnectorOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/connector",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams()
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(List)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new connector.
        /// </summary>
        public async Task<ConnectorOut> CreateAsync(
            ConnectorIn connectorIn,
            ConnectorCreateOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            connectorIn = connectorIn ?? throw new ArgumentNullException(nameof(connectorIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ConnectorOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/connector",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: connectorIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(CreateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new connector.
        /// </summary>
        public ConnectorOut Create(ConnectorIn connectorIn, ConnectorCreateOptions? options = null)
        {
            connectorIn = connectorIn ?? throw new ArgumentNullException(nameof(connectorIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ConnectorOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/connector",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: connectorIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Create)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get a connector.
        /// </summary>
        public async Task<ConnectorOut> GetAsync(
            string connectorId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ConnectorOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(GetAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get a connector.
        /// </summary>
        public ConnectorOut Get(string connectorId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ConnectorOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Get)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update a connector.
        /// </summary>
        public async Task<ConnectorOut> UpdateAsync(
            string connectorId,
            ConnectorUpdate connectorUpdate,
            CancellationToken cancellationToken = default
        )
        {
            connectorUpdate =
                connectorUpdate ?? throw new ArgumentNullException(nameof(connectorUpdate));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ConnectorOut>(
                    method: HttpMethod.Put,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    content: connectorUpdate,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(UpdateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update a connector.
        /// </summary>
        public ConnectorOut Update(string connectorId, ConnectorUpdate connectorUpdate)
        {
            connectorUpdate =
                connectorUpdate ?? throw new ArgumentNullException(nameof(connectorUpdate));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ConnectorOut>(
                    method: HttpMethod.Put,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    content: connectorUpdate
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Update)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete a connector.
        /// </summary>
        public async Task<bool> DeleteAsync(
            string connectorId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<bool>(
                    method: HttpMethod.Delete,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(DeleteAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete a connector.
        /// </summary>
        public bool Delete(string connectorId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<bool>(
                    method: HttpMethod.Delete,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Delete)} failed");

                throw;
            }
        }

        /// <summary>
        /// Partially update a connector.
        /// </summary>
        public async Task<ConnectorOut> PatchAsync(
            string connectorId,
            ConnectorPatch connectorPatch,
            CancellationToken cancellationToken = default
        )
        {
            connectorPatch =
                connectorPatch ?? throw new ArgumentNullException(nameof(connectorPatch));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ConnectorOut>(
                    method: HttpMethod.Patch,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    content: connectorPatch,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(PatchAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Partially update a connector.
        /// </summary>
        public ConnectorOut Patch(string connectorId, ConnectorPatch connectorPatch)
        {
            connectorPatch =
                connectorPatch ?? throw new ArgumentNullException(nameof(connectorPatch));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ConnectorOut>(
                    method: HttpMethod.Patch,
                    path: "/api/v1/connector/{connector_id}",
                    pathParams: new Dictionary<string, string> { { "connector_id", connectorId } },
                    content: connectorPatch
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Patch)} failed");

                throw;
            }
        }
    }
}

File: svix-cli/src/cmds/api/connector.rs
// this file is @generated
use clap::{Args, Subcommand};
use svix::api::Svix;
#[allow(unused_imports)]
use svix::models::*;

#[derive(Args, Clone)]
pub struct ConnectorListOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// The sorting order of the returned items
    #[arg(long)]
    pub order: Option<Ordering>,
    #[arg(long)]
    pub product_type: Option<ConnectorProduct>,
}

impl From<ConnectorListOptions> for svix::api::ConnectorListOptions {
    fn from(value: ConnectorListOptions) -> Self {
        let ConnectorListOptions {
            limit,
            iterator,
            order,
            product_type,
        } = value;
        Self {
            limit,
            iterator,
            order,
            product_type,
        }
    }
}

#[derive(Args, Clone)]
pub struct ConnectorCreateOptions {
    #[arg(long)]
    pub idempotency_key: Option<String>,
}

impl From<ConnectorCreateOptions> for svix::api::ConnectorCreateOptions {
    fn from(value: ConnectorCreateOptions) -> Self {
        let ConnectorCreateOptions { idempotency_key } = value;
        Self { idempotency_key }
    }
}

#[derive(Args)]
#[command(args_conflicts_with_subcommands = true, flatten_help = true)]
pub struct ConnectorArgs {
    #[command(subcommand)]
    pub command: ConnectorCommands,
}

#[allow(clippy::large_enum_variant)]
#[derive(Subcommand)]
pub enum ConnectorCommands {
    /// List all connectors for an application.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector list\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    #[command(after_help = "Example response:
{
  \"data\": [{
    \"id\": \"trtmpl_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
    \"orgId\": \"org_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
    \"uid\": \"unique-identifier\",
    \"kind\": \"Custom\",
    \"name\": \"...\",
    \"logo\": \"...\",
    \"description\": \"...\",
    \"instructions\": \"...\",
    \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
    \"transformation\": \"...\",
    \"createdAt\": \"2030-01-01T00:00:00Z\",
    \"updatedAt\": \"2030-01-01T00:00:00Z\",
    \"transformationUpdatedAt\": \"2030-01-01T00:00:00Z\",
    \"featureFlags\": [\"cool-new-feature\"],
    \"productType\": \"Dispatch\"
  }],
  \"iterator\": \"iterator\",
  \"prevIterator\": \"-iterator\",
  \"done\": true
}\n")]
    List {
        #[clap(flatten)]
        options: ConnectorListOptions,
    },
    /// Create a new connector.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector create {...}\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    #[command(after_help = "Example body:
{
  \"name\": \"My first connector\",
  \"uid\": \"unique-identifier\",
  \"logo\": \"https://example.com/logo.png\",
  \"description\": \"Example connector description\",
  \"kind\": \"Slack\",
  \"instructions\": \"Markdown-formatted instructions\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"function handler(webhook) { /* ... */ }\",
  \"featureFlags\": [\"...\"],
  \"productType\": \"Dispatch\"
}\n\nExample response:
{
  \"id\": \"trtmpl_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"orgId\": \"org_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"uid\": \"unique-identifier\",
  \"kind\": \"Custom\",
  \"name\": \"...\",
  \"logo\": \"...\",
  \"description\": \"...\",
  \"instructions\": \"...\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"...\",
  \"createdAt\": \"2030-01-01T00:00:00Z\",
  \"updatedAt\": \"2030-01-01T00:00:00Z\",
  \"transformationUpdatedAt\": \"2030-01-01T00:00:00Z\",
  \"featureFlags\": [\"cool-new-feature\"],
  \"productType\": \"Dispatch\"
}\n")]
    Create {
        connector_in: crate::json::JsonOf<ConnectorIn>,
        #[clap(flatten)]
        options: ConnectorCreateOptions,
    },
    /// Get a connector.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector get CONNECTOR_ID\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    #[command(after_help = "Example response:
{
  \"id\": \"trtmpl_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"orgId\": \"org_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"uid\": \"unique-identifier\",
  \"kind\": \"Custom\",
  \"name\": \"...\",
  \"logo\": \"...\",
  \"description\": \"...\",
  \"instructions\": \"...\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"...\",
  \"createdAt\": \"2030-01-01T00:00:00Z\",
  \"updatedAt\": \"2030-01-01T00:00:00Z\",
  \"transformationUpdatedAt\": \"2030-01-01T00:00:00Z\",
  \"featureFlags\": [\"cool-new-feature\"],
  \"productType\": \"Dispatch\"
}\n")]
    Get { id: String },
    /// Create or update a connector.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector upsert CONNECTOR_ID {...}\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    #[command(after_help = "Example body:
{
  \"name\": \"My first connector\",
  \"logo\": \"https://example.com/logo.png\",
  \"description\": \"Example connector description\",
  \"kind\": \"Slack\",
  \"instructions\": \"Markdown-formatted instructions\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"function handler(webhook) { /* ... */ }\",
  \"featureFlags\": [\"cool-new-feature\"]
}\n\nExample response:
{
  \"id\": \"trtmpl_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"orgId\": \"org_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"uid\": \"unique-identifier\",
  \"kind\": \"Custom\",
  \"name\": \"...\",
  \"logo\": \"...\",
  \"description\": \"...\",
  \"instructions\": \"...\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"...\",
  \"createdAt\": \"2030-01-01T00:00:00Z\",
  \"updatedAt\": \"2030-01-01T00:00:00Z\",
  \"transformationUpdatedAt\": \"2030-01-01T00:00:00Z\",
  \"featureFlags\": [\"cool-new-feature\"],
  \"productType\": \"Dispatch\"
}\n")]
    Upsert {
        id: String,
        connector_update: crate::json::JsonOf<ConnectorUpdate>,
    },
    /// Delete a connector.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector delete CONNECTOR_ID\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    Delete { id: String },
    /// Partially update a connector.
    #[command(help_template = concat!(
            "{about-with-newline}\n",
            "{usage-heading} {usage}\n\n",
            "Example: svix connector patch CONNECTOR_ID {...}\n",
            "{after-help}",
            "\n",
            "{all-args}",
        ))]
    #[command(after_help = "Example body:
{
  \"name\": \"...\",
  \"logo\": \"...\",
  \"description\": \"...\",
  \"kind\": \"Custom\",
  \"instructions\": \"...\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"...\",
  \"featureFlags\": [\"cool-new-feature\"]
}\n\nExample response:
{
  \"id\": \"trtmpl_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"orgId\": \"org_1srOrx2ZWZBpBUvZwXKQmoEYga2\",
  \"uid\": \"unique-identifier\",
  \"kind\": \"Custom\",
  \"name\": \"...\",
  \"logo\": \"...\",
  \"description\": \"...\",
  \"instructions\": \"...\",
  \"allowedEventTypes\": [\"user.signup\",\"user.deleted\"],
  \"transformation\": \"...\",
  \"createdAt\": \"2030-01-01T00:00:00Z\",
  \"updatedAt\": \"2030-01-01T00:00:00Z\",
  \"transformationUpdatedAt\": \"2030-01-01T00:00:00Z\",
  \"featureFlags\": [\"cool-new-feature\"],
  \"productType\": \"Dispatch\"
}\n")]
    Patch {
        id: String,
        connector_patch: Option<crate::json::JsonOf<ConnectorPatch>>,
    },
}

impl ConnectorCommands {
    pub async fn exec(
        self,
        client: &Svix,
        color_mode: colored_json::ColorMode,
    ) -> anyhow::Result<()> {
        match self {
            Self::List { options } => {
                let resp = client.connector().list(Some(options.into())).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Create {
                connector_in,
                options,
            } => {
                let resp = client
                    .connector()
                    .create(connector_in.into_inner(), Some(options.into()))
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Get { id } => {
                let resp = client.connector().get(id).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Upsert {
                id,
                connector_update,
            } => {
                let resp = client
                    .connector()
                    .upsert(id, connector_update.into_inner())
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Delete { id } => {
                client.connector().delete(id).await?;
            }
            Self::Patch {
                id,
                connector_patch,
            } => {
                let resp = client
                    .connector()
                    .patch(id, connector_patch.unwrap_or_default().into_inner())
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
        }

        Ok(())
    }
}
